Add beginner documentation (#397)
bats-core/bats-core: docs/source/tutorial.rst

    } 

    @test "Show welcome message on first invocation" {
        if [[ -e /tmp/bats-tutorial-project-ran ]]; then
            skip 'The FIRST_RUN_FILE already exists'
        fi
        
        run project.sh
        assert_output --partial 'Welcome to our project!'

        run project.sh
        refute_output --partial 'Welcome to our project!'
    }

The first test run still works due to the cleanup from the last round. However, our second run gives us:

.. code-block:: console

    $ ./test/bats/bin/bats test/test.bats 
     - Show welcome message on first invocation (skipped: The FIRST_RUN_FILE already exists)

    1 test, 0 failures, 1 skipped

.. important::

    Skipped tests won't fail a test suite and are counted separately.
    No test command after `skip` will be executed. If an error occurs before `skip`, the test will fail.
    An optional reason can be passed to `skip` and will be printed in the test output.

Setting up a multifile test suite
---------------------------------

With a growing project, putting all tests into one file becomes unwieldy.
For our example project, we will extract functionality into the additional file `src/helper.sh`:

> ,, code-block:: bash

Reviewer: Typo, should be `..`

#!/usr/bin/env bash

    _is_first_run() {
        local FIRST_RUN_FILE=/tmp/bats-tutorial-project-ran
        if [[ ! -e "$FIRST_RUN_FILE" ]]; then
            touch "$FIRST_RUN_FILE"
            return 0
        fi
        return 1
    }

This allows for testing it separately in a new file `test/helper.bats`:

> .. code-block:: bash

Reviewer: Annoyingly this code highlighting doesn't seem to work, is it the shebang that does it? The next block is ok, for example

Author: There was an error in the code which probably derailed the parser.

Tutorial
========

This tutorial is intended for beginners with bats and possibly bash.
Make sure to also read the list of gotchas and the faq.

For this tutorial we are assuming you already have a project in a git repository and want to add tests.

> Ultimately they should run in the CI environment but will also be started locally during development.

Reviewer 2: > Ultimately they should run in the CI environment but will also be started locally during development.

This is a subjective thing - I believe that particular sentence is a bit out of place since the writing of the tests will be the same process whether or not it runs locally or in CI. Maybe we can add a separate TODO section for "Example of bats in CI"?

> Ultimately they should run in the CI environment but will also be started locally during development.

Author: What I meant to express here is: Most editing/test development won't be done on the CI system. But automated tests are mainly intended to be run in a CI environment.
Of course, there are other workflows but in my view there are two main usecases: 

1. unsupervised batch execution, e.g. on the CI server
2. interactive execution, e.g. on the developer machine

Obviously, there is are cases in between, like `git bisect run`.

        # use $BATS_TEST_FILENAME instead of ${BASH_SOURCE[0]} or $0,
        # as those will point to the bats executable's location or the preprocessed file respectively
        DIR="$( cd "$( dirname "$BATS_TEST_FILENAME" )" >/dev/null 2>&1 && pwd )"
        # make executables in src/ visible to PATH
        PATH="$DIR/../src:$PATH"
    }

    @test "can run our script" {
        run project.sh # notice `run`!
        assert_output 'Welcome to our project!'
    }

which gives us the following test output:

.. code-block:: console

    $ LANG=C ./test/bats/bin/bats test/test.bats 
     ✗ can run our script
       (from function `assert_output' in file test/test_helper/bats-assert/src/assert_output.bash, line 194,
        in test file test/test.bats, line 14)
         `assert_output 'Welcome to our project!'' failed
    
       -- output differs --
       expected (1 lines):
         Welcome to our project!
       actual (2 lines):
         Welcome to our project!
         NOT IMPLEMENTED!
       --
    

    1 test, 1 failure

The first change in this output is the failure description. We now fail on assert_output instead of the call itself.

> We prefixed our call to `project.sh` with `run`, which is an internal bats function that executes the command it gets passed as parameters.

Reviewer 2: ```suggestion
We prefixed our call to `project.sh` with `run`, which is a bats function that executes the command it gets passed as parameters.
```

Since we expect users to use it, I don't think it should be described as an internal function.

Then, `run` sucks up the stdout and stderr of the command it ran and stores it in `$output`, stores the exit code in `$status` and returns 0.
This means `run` never fails the test and won't generate any context/output in the log of a failed test on its own.

> Failing the test is and printing context information is up to the consumers of `$status` and `$output`. 

Reviewer 2: > Failing the test is and printing context information is up to the consumers of `$status` and `$output`. 

A bit confused by the first part of this sentence - can we rephrase it?

Author: I hope it is a bit clearer now.

`grep` is always at our fingertips, so we tape together this ramshackle construct

.. code-block:: bash

    run project.sh 2>&1 | grep Welcome

which gives us the following test result:

.. code-block:: console

    $ LANG=C ./test/bats/bin/bats test/test.bats 
     ✗ can run our script
       (in test file test/test.bats, line 13)
         `run project.sh | grep Welcome' failed

    1 test, 1 failure

Huh, what is going on? Why does it fail the `run` line again?

This is a common mistake that can happen when our mind parses the file differently than the bash parser.
`run` is just a function, so the pipe won't actually be forwarded into the function. Bash reads this as `(run project.sh) | grep Welcome`, 
instead of our intended `run (project.sh | grep Welcome)`.

Unfortunately, the latter is no valid bash syntax, so we have to work around it, e.g. by using a function:

.. code-block:: bash

    get_projectsh_welcome_message() {
        project.sh  2>&1 | grep Welcome
    }

    @test "Check welcome message" {
        run get_projectsh_welcome_message
        assert_output 'Welcome to our project!'
    }

> Now our test passes again but having to write a function each time we want only a partial match does not accomodate our lazyness.

Reviewer 2: ```suggestion
Now our test passes again but having to write a function each time we want only a partial match does not accommodate our laziness.
```

Nice, we're done, or are we? Running the test again now gives:

.. code-block:: console

    $ ./test/bats/bin/bats test/test.bats 
     ✗ Show welcome message on first invocation
       (from function `assert_output' in file test/test_helper/bats-assert/src/assert_output.bash, line 186,
        in test file test/test.bats, line 14)
         `assert_output --partial 'Welcome to our project!'' failed
    
       -- output does not contain substring --
       substring : Welcome to our project!
       output    : NOT IMPLEMENTED!
       --
    

    1 test, 1 failure

Now the first assert failed, because of the leftover `$FIRST_RUN_FILE` from the last test run.

Luckily, bats offers the `teardown` function, which can take care of that, we add the following code to `test/test.bats`:

.. code-block:: bash

    teardown() {
        rm -f /tmp/bats-tutorial-project-ran
    }

Now running the test again first give us the same error, as the teardown has not run yet. 
On the second try we get a clean `/tmp` folder again and our test passes consistently now.

It is worth noting that we could do this `rm` in the test code itself but it would get skipped on failures.

.. important::

>     A test ends at its first failure. Non of the following commands in this test will be executed.

Reviewer 2: ```suggestion
    A test ends at its first failure. None of the subsequent commands in this test will be executed.
```
